iml-services/iml-devices/src/db.rs

```diff
@@ -0,0 +1,482 @@
+        mount_path: MountPath(flat_dev.mount_path.clone()),
+        fs_type: flat_dev.fs_type.clone(),
+        fs_label: flat_dev.fs_label.clone(),
+        fs_uuid: flat_dev.fs_uuid.clone(),
+    };
+
+    (
+        (flat_dev.id.clone(), d),
+        ((flat_dev.id.clone(), dh.fqdn.clone()), dh),
+    )
+}
+
+pub fn convert_flat_devices(flat_devices: &FlatDevices, fqdn: Fqdn) -> (Devices, DeviceHosts) {
+    flat_devices
+        .iter()
+        .map(|x| create_dev(x.1, fqdn.clone()))
+        .unzip()
+}
+
+/// Given a device id and some `DeviceHosts`,
+/// try to find the first cooresponding host.
+fn find_device_host<'a>(
+    id: &'a DeviceId,
+    device_hosts: &'a DeviceHostsRef<'a>,
+) -> Option<(&'a &'a DeviceHostKey, &'a &'a DeviceHost)> {
+    device_hosts
+        .into_iter()
+        .find(move |(_, v)| &v.device_id == id)
+}
+
+pub fn get_local_device_hosts<'a>(
+    device_hosts: &'a DeviceHosts,
+    fqdn: &Fqdn,
+) -> DeviceHostsRef<'a> {
+    device_hosts
+        .into_iter()
+        .filter(|(_k, v)| &v.fqdn == fqdn && v.local)
+        .collect()
+}
+
+pub fn get_local_devices<'a>(
+    local_device_hosts: &DeviceHostsRef<'_>,
+    devices: &'a Devices,
+) -> DevicesRef<'a> {
+    devices
+        .into_iter()
+        .filter(|(k, _)| {
+            local_device_hosts
+                .into_iter()
+                .find(|(_, v)| &&v.device_id == k)
+                .is_some()
+        })
+        .collect()
+}
+
+pub fn get_other_device_hosts<'a>(
+    db_device_hosts: &'a DeviceHosts,
+    fqdn: &'a Fqdn,
+) -> DeviceHostsRef<'a> {
+    db_device_hosts
+        .iter()
+        .filter(move |(_, v)| &v.fqdn != fqdn)
+        .collect()
+}
+
+pub fn get_devices_by_device_host<'a>(
+    device_hosts: &'a DeviceHostsRef<'a>,
+    devices: &'a Devices,
+) -> DevicesRef<'a> {
+    devices
+        .iter()
+        .filter(move |(k, _)| find_device_host(&k, device_hosts).is_some())
+        .collect()
+}
+
+pub async fn get_db_devices(mut client: &mut Client) -> Result<Devices, iml_postgres::Error> {
+    select_all(
+        &mut client,
+        &format!("SELECT * FROM {}", Device::table_name()),
+        iter::empty(),
+    )
+    .await?
+    .map_ok(Device::from)
+    .map_ok(|x| (x.id.clone(), x))
+    .try_collect()
+    .await
+}
+
+pub async fn get_db_device_hosts(
+    mut client: &mut Client,
+) -> Result<Vec<DeviceHost>, iml_postgres::Error> {
+    select_all(
+        &mut client,
+        &format!("SELECT * FROM {}", DeviceHost::table_name()),
+        iter::empty(),
+    )
+    .await?
+    .map_ok(DeviceHost::from)
+    .try_collect()
+    .await
+}
+
+async fn insert_device_host(
+    transaction: &mut Transaction<'_>,
+    fqdn: &Fqdn,
+    x: &DeviceHost,
+) -> Result<(), ImlDevicesError> {
+    let s = transaction.prepare(
+        &format!("INSERT INTO {} (device_id, fqdn, local, paths, mount_path, fs_type, fs_label, fs_uuid) VALUES ($1, $2, $3, $4, $5, $6, $7, $8)", DeviceHost::table_name())
```

Comment:
It looks like there are no benefits from preparing a statement, because it will be executed only once.
And thus, the code could be shorter by providing the query string [directly to the `execute()` method](https://docs.rs/tokio-postgres/0.5.2/tokio_postgres/struct.Client.html#method.execute).

> A statement may contain parameters, specified by `$n`, where n is the index of the parameter of the list provided, 1-indexed. The statement argument can either be a Statement, or a raw query string. If the same statement will be repeatedly executed (perhaps with different query parameters), consider preparing the statement up front with the `prepare` method.

Reply:
We're preparing the statement separately in all of the codebase.

Even if not for consistency, I guess the reasoning could be that two shorter async locking operations might be better than one longer one, since it will have more interrupt points for other futures depending on same resource.

There's also [pipelining](https://docs.rs/tokio-postgres/0.5.2/tokio_postgres/#pipelining) to consider.

Reply:
It is still async under the hood.

Reply:
Will change to `execute` as decided on call.

Reply:
Done